File: apps/mobile/shared/Sources/FlagshipCore/ServerTransferFlow.swift
import Foundation
import CryptoKit
import Security
import FlagshipAPI

/// Pure (testable) builders for the transfer-a-box flow
/// (docs/account-deletion-and-name-reclaim.md §4). The SwiftUI VMs derive the
/// IRK behind the biometric, then call these to produce the exact wire bodies
/// the broker accepts — byte-identical to the webapp `lib/serverTransfer.js`.
///
/// Split this way so the crypto/canonical-bytes is `swift test`-able without the
/// UIKit-bound VM layer.
public enum ServerTransferFlow {

    /// The QR payload the giver shows + the acquirer scans. JSON, byte-stable
    /// field set matching the webapp `createTransferOffer().qr`.
    public struct OfferQR: Codable, Equatable, Sendable {
        public let v: Int
        public let kind: String
        public let serverDomain: String
        public let transferNonce: String
        public let giverIrkPub: String
        public let issuedAt: Int64
        public let expiresAt: Int64
        public let offerSignature: String

        public init(serverDomain: String, transferNonce: String, giverIrkPub: String, issuedAt: Int64, expiresAt: Int64, offerSignature: String) {
            self.v = 1
            self.kind = "flagship-transfer-offer"
            self.serverDomain = serverDomain
            self.transferNonce = transferNonce
            self.giverIrkPub = giverIrkPub
            self.issuedAt = issuedAt
            self.expiresAt = expiresAt
            self.offerSignature = offerSignature
        }
    }

    public enum TransferError: Error, Equatable, Sendable {
        case notATransferQR
        case malformedQR
        case expired
        /// The offer's `offerSignature` does not verify under `giverIrkPub` over
        /// the `ServerTransferOfferOrder` canonical bytes. A deep-linked/scanned
        /// offer is attacker-supplied — this is the forgery gate.
        case badSignature
    }

    // MARK: - Universal-link / custom-scheme QR encoding (Slice C)

    /// The path the transfer universal link + custom scheme both use.
    public static let transferLinkPath = "/transfer"
    /// The query param carrying the base64url(offerJSON).
    public static let transferLinkParam = "o"

    /// Encode an offer as the `o=` param: `base64url(UTF8(offerJSON))`, NO padding
    /// (`-_` alphabet, `=` stripped). `offerJSON` is `encodeQR` (sortedKeys), so
    /// the payload is byte-identical to what the in-app scanner parses.
    public static func encodeOfferParam(_ qr: OfferQR) throws -> String {
        let json = try encodeQR(qr)
        return base64URLNoPadding(Data(json.utf8))
    }

    /// Decode an `o=` param back to the offer JSON string. Returns nil on a
    /// malformed base64url / non-UTF8 payload.
    public static func decodeOfferParam(_ b64url: String) -> String? {
        guard let data = decodeBase64URLNoPadding(b64url),
              let json = String(data: data, encoding: .utf8) else { return nil }
        return json
    }

    /// The UNIVERSAL LINK the giver renders as a QR so the native Camera can open
    /// it: `https://flagshipserver.com/transfer?o=<b64url>`. Uses the configured
    /// control host (prod = `flagshipserver.com`).
    public static func transferUniversalLink(_ qr: OfferQR, controlHost: String = Endpoints.controlHost) throws -> String {
        "https://\(controlHost)\(transferLinkPath)?\(transferLinkParam)=\(try encodeOfferParam(qr))"
    }

    /// The custom-scheme twin: `flagship://transfer?o=<b64url>`.
    public static func transferCustomSchemeLink(_ qr: OfferQR) throws -> String {
        "flagship://transfer?\(transferLinkParam)=\(try encodeOfferParam(qr))"
    }

    /// Base64url-no-padding encoder (RFC 4648 §5) — matches the webapp/Android
    /// encoders + the existing `CompanionTicketURL` idiom.
    static func base64URLNoPadding(_ data: Data) -> String {
        var s = data.base64EncodedString()
        s = s.replacingOccurrences(of: "+", with: "-")
        s = s.replacingOccurrences(of: "/", with: "_")
        while s.hasSuffix("=") { s.removeLast() }
        return s
    }

    static func decodeBase64URLNoPadding(_ s: String) -> Data? {
        var t = s
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while t.count % 4 != 0 { t.append("=") }
        return Data(base64Encoded: t)
    }

    // MARK: - GIVER: build the offer body + the QR

    /// Sign a `ServerTransferOffer` + an IRK mailbox-auth, returning the deposit
    /// body AND the QR to render. `ttlMs` defaults to 15 min (the broker clamps).
    public static func buildOffer(
        serverDomain: String,
        username: String,
        irk: Curve25519.Signing.PrivateKey,
        orderKey: Curve25519.Signing.PrivateKey? = nil,
        issuedAt: Int64,
        ttlMs: Int64 = 15 * 60_000,
        nonce: Data,
        authNonce: Data
    ) throws -> (body: TransferOfferBody, qr: OfferQR) {
        let nonceHex = HexUtil.encode(nonce)
        let expiresAt = issuedAt + ttlMs
        let order = ServerTransferOfferOrder(
            serverDomain: serverDomain, transferNonce: nonceHex, issuedAt: issuedAt, expiresAt: expiresAt
        )
        // Slice D — the transfer OFFER is a SENSITIVE order: sign with the giver's
        // admin master root (`orderKey`) when supplied, else the IRK. The QR's
        // `giverIrkPub` MUST be the key the offer was signed with (the acquirer's
        // local `verifyOffer` checks `offerSignature` under it), so it tracks the
        // signing key. `.com` gates the offer against the giver's admin root and
        // records the giver identity from the registered account, independent of
        // this QR field. The mailbox AUTH stays IRK-signed.
        let signKey = orderKey ?? irk
        let offerSig = try order.sign(with: signKey)
        let offerSigHex = HexUtil.encode(offerSig)
        let auth = try buildMailboxAuth(username: username, irk: irk, issuedAt: issuedAt, nonce: authNonce)
        let body = TransferOfferBody(
            auth: auth.auth,
            authSignature: auth.authSignature,
            offer: TransferOfferWire(serverDomain: serverDomain, transferNonce: nonceHex, issuedAt: issuedAt, expiresAt: expiresAt),
            offerSignature: offerSigHex
        )
        let qr = OfferQR(
            serverDomain: serverDomain,
            transferNonce: nonceHex,
            giverIrkPub: HexUtil.encode(signKey.publicKey.rawRepresentation),
            issuedAt: issuedAt,
            expiresAt: expiresAt,
            offerSignature: offerSigHex
        )
        return (body, qr)
    }

    /// Cheap shape check for the camera validator (full parse happens on scan).
    /// Accepts BOTH the raw offer JSON (carries the `flagship-transfer-offer`
    /// kind) AND the universal-link / custom-scheme QR (`…/transfer?o=`).
    public static func looksLikeTransferQR(_ text: String) -> Bool {
        text.contains("flagship-transfer-offer")
            || text.contains("\(transferLinkPath)?\(transferLinkParam)=")
    }

    public static func encodeQR(_ qr: OfferQR) throws -> String {
        let enc = JSONEncoder()
        enc.outputFormatting = [.sortedKeys]
        return String(data: try enc.encode(qr), encoding: .utf8) ?? ""
    }

    // MARK: - ACQUIRER: parse + build the claim body

    /// Parse a scanned/pasted QR string. Throws on a malformed / wrong-kind payload.
    public static func parseQR(_ text: String) throws -> OfferQR {
        guard let data = text.data(using: .utf8) else { throw TransferError.notATransferQR }
        guard let qr = try? JSONDecoder().decode(OfferQR.self, from: data) else {
            throw TransferError.notATransferQR
        }
        guard qr.kind == "flagship-transfer-offer" else { throw TransferError.notATransferQR }
        guard !qr.serverDomain.isEmpty, !qr.transferNonce.isEmpty, !qr.giverIrkPub.isEmpty else {
            throw TransferError.malformedQR
        }
        return qr
    }

    /// Tolerant parse for a scanned string that may be EITHER the raw offer JSON
    /// OR a `flagship://transfer?o=` / `https://<controlHost>/transfer?o=` link.
    /// The in-app scanner uses this so it keeps working whether the giver renders
    /// the JSON (older client) or the universal link (Slice C).
    public static func parseScanned(_ text: String) throws -> OfferQR {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if let json = offerJSONFromLink(trimmed) {
            return try parseQR(json)
        }
        return try parseQR(trimmed)
    }

    /// Pull the offer JSON out of a transfer link (universal or custom scheme),
    /// or nil if the string isn't a `…/transfer?o=…` link.
    public static func offerJSONFromLink(_ text: String) -> String? {
        guard let url = URL(string: text),
              let comps = URLComponents(url: url, resolvingAgainstBaseURL: false),
              comps.path == transferLinkPath || url.host == "transfer",
              let o = comps.queryItems?.first(where: { $0.name == transferLinkParam })?.value
        else { return nil }
        return decodeOfferParam(o)
    }

    /// SECURITY GATE (Slice C) — Ed25519-verify `offer.offerSignature` under
    /// `offer.giverIrkPub` over the `ServerTransferOfferOrder` canonical bytes AND
    /// require `expiresAt > now`. A deep-linked/scanned offer is attacker-supplied,
    /// so the acquirer MUST pass this before building any claim. Throws
    /// `.malformedQR` (bad pubkey), `.badSignature`, or `.expired`.
    public static func verifyOffer(_ offer: OfferQR, now: Int64) throws {
        guard let pubData = HexUtil.decode(offer.giverIrkPub), pubData.count == 32,
              let pub = try? Curve25519.Signing.PublicKey(rawRepresentation: pubData)
        else { throw TransferError.malformedQR }
        guard let sig = HexUtil.decode(offer.offerSignature) else { throw TransferError.badSignature }
        let order = ServerTransferOfferOrder(
            serverDomain: offer.serverDomain,
            transferNonce: offer.transferNonce,
            issuedAt: offer.issuedAt,
            expiresAt: offer.expiresAt
        )
        guard pub.isValidSignature(sig, for: order.canonicalBytes()) else {
            throw TransferError.badSignature
        }
        guard offer.expiresAt > now else { throw TransferError.expired }
    }

    /// Sign a `ServerTransferClaim` for a parsed offer, returning the claim body.
    public static func buildClaim(
        offer: OfferQR,
        acquirerUsername: String,
        acquirerIrk: Curve25519.Signing.PrivateKey,
        orderKey: Curve25519.Signing.PrivateKey? = nil,
        acquirerAdminRootPubHex: String = "",
        issuedAt: Int64
    ) throws -> TransferClaimBody {
        if offer.expiresAt <= issuedAt { throw TransferError.expired }
        // The claim's `acquirerIrkPubHex` field STAYS the acquirer's registered
        // IRK — `.com` requires `claim.acquirerIrkPub == acquirer.irkPubHex`
        // (identity), independent of the signature. Slice D signs the SENSITIVE
        // claim order with the acquirer's admin master root (`orderKey`) when
        // supplied (else the IRK); `.com` gates the signature against the
        // acquirer's admin root. §9.8: the v2 canonical additionally commits to
        // `acquirerAdminRootPubHex` ("" when the account has none) — the anchor
        // the box re-pins on re-home rides INSIDE the acquirer's signature, so
        // `.com` can't substitute it.
        let acquirerIrkHex = HexUtil.encode(acquirerIrk.publicKey.rawRepresentation)
        let lowered = acquirerUsername.lowercased()
        let order = ServerTransferClaimOrder(
            serverDomain: offer.serverDomain,
            transferNonce: offer.transferNonce,
            acquirerUsername: lowered,
            acquirerIrkPubHex: acquirerIrkHex,
            acquirerAdminRootPubHex: acquirerAdminRootPubHex,
            issuedAt: issuedAt
        )
        let sig = try order.sign(with: orderKey ?? acquirerIrk)
        return TransferClaimBody(
            claim: TransferClaimWire(
                serverDomain: offer.serverDomain,
                transferNonce: offer.transferNonce,
                acquirerUsername: lowered,
                acquirerIrkPub: acquirerIrkHex,
                acquirerAdminRootPub: acquirerAdminRootPubHex,
                issuedAt: issuedAt
            ),
            claimSignature: HexUtil.encode(sig)
        )
    }

    // MARK: - GIVER: disk-key re-seal

    /// Re-seal the box's LUKS disk key (already unsealed by the giver IRK) to the
    /// ACQUIRER IRK pub, returning the deposit body. `.com` stays content-blind.
    public static func buildDiskKeyDeposit(
        serverDomain: String,
        username: String,
        irk: Curve25519.Signing.PrivateKey,
        diskKey: Data,
        acquirerIrkPubHex: String,
        issuedAt: Int64,
        authNonce: Data
    ) throws -> TransferDiskKeyBody {
        guard let acquirerPub = HexUtil.decode(acquirerIrkPubHex), acquirerPub.count == 32 else {
            throw TransferError.malformedQR
        }
        let sealed = try SecretSeal.sealForEd25519Recipient(plaintext: diskKey, recipientEd25519Pub: acquirerPub)
        let auth = try buildMailboxAuth(username: username, irk: irk, issuedAt: issuedAt, nonce: authNonce)
        return TransferDiskKeyBody(
            auth: auth.auth,
            authSignature: auth.authSignature,
            sealedDiskKey: HexUtil.encode(sealed)
        )
    }

    /// ACQUIRER: open the giver's re-sealed disk key with the acquirer IRK seed.
    public static func openDiskKey(sealedHex: String, acquirerIrk: Curve25519.Signing.PrivateKey) throws -> Data {
        guard let blob = HexUtil.decode(sealedHex) else { throw TransferError.malformedQR }
        return try SecretSeal.openWithEd25519Seed(blob: blob, recipientEd25519Seed: acquirerIrk.rawRepresentation)
    }

    // MARK: - GIVER: admin-root hand-off (spec §9.8)

    /// Build + sign the `flagship/admin-root-transfer/v1` hand-off proof with
    /// the GIVER's admin master root (the box's pinned anchor). The box
    /// verifies the signature against its pin before re-pinning to
    /// `acquirerAdminRootPubHex` — "" means unpin (acquirer has no admin root).
    /// `serverDomain` is the box's OLD canonical.
    public static func buildAdminHandoff(
        serverDomain: String,
        giverUsername: String,
        acquirerUsername: String,
        acquirerAdminRootPubHex: String,
        giverAdminRoot: Curve25519.Signing.PrivateKey,
        transferNonce: String,
        issuedAt: Int64
    ) throws -> TransferAdminHandoffBody {
        let oldPubHex = HexUtil.encode(giverAdminRoot.publicKey.rawRepresentation)
        let handoff = AdminRootTransfer(
            serverDomain: serverDomain,
            giverUsername: giverUsername,
            acquirerUsername: acquirerUsername,
            oldAdminRootPubHex: oldPubHex,
            newAdminRootPubHex: acquirerAdminRootPubHex,
            transferNonce: transferNonce,
            issuedAt: issuedAt
        )
        let sig = try handoff.sign(withGiverAdminRoot: giverAdminRoot)
        return TransferAdminHandoffBody(
            handoff: TransferAdminHandoffWire(
                serverDomain: serverDomain,
                giverUsername: giverUsername,
                acquirerUsername: acquirerUsername,
                oldAdminRootPub: oldPubHex,
                newAdminRootPub: acquirerAdminRootPubHex,
                transferNonce: transferNonce,
                issuedAt: issuedAt
            ),
            signatureHex: HexUtil.encode(sig)
        )
    }

    // MARK: - GIVER: legacy re-home authorization (v1-sec GAP 3)

    /// Build + sign the `flagship/server-rehome-auth/v1` re-home authorization
    /// with the GIVER's owner IRK (the box's pinned owner IRK until it re-homes).
    /// A box with NO pinned admin master root verifies this against its pin
    /// before writing the re-home marker — never on `.com`'s unsigned word. The
    /// deposit body carries only `issuedAt` + the signature; `.com` reconstructs
    /// the signed (old/new domain, acquirer IRK) fields from the claimed row.
    /// `oldServerDomain` is the box's OLD canonical; `newServerDomain` +
    /// `acquirerIrkPubHex` come from the giver's claim poll.
    public static func buildRehomeAuth(
        oldServerDomain: String,
        newServerDomain: String,
        acquirerIrkPubHex: String,
        giverIrk: Curve25519.Signing.PrivateKey,
        issuedAt: Int64
    ) throws -> TransferRehomeAuthBody {
        let order = RehomeAuthorizationOrder(
            oldServerDomain: oldServerDomain,
            newServerDomain: newServerDomain,
            acquirerIrkPubHex: acquirerIrkPubHex,
            issuedAt: issuedAt
        )
        let sig = try order.sign(withGiverIrk: giverIrk)
        return TransferRehomeAuthBody(issuedAt: issuedAt, signatureHex: HexUtil.encode(sig))
    }

    /// GIVER: open the box's install-time disk key (sealed FOR the giver IRK)
    /// with the giver IRK seed — the first half of the re-seal step.
    public static func openGiverDiskKey(sealedHex: String, giverIrk: Curve25519.Signing.PrivateKey) throws -> Data {
        guard let blob = HexUtil.decode(sealedHex) else { throw TransferError.malformedQR }
        return try SecretSeal.openWithEd25519Seed(blob: blob, recipientEd25519Seed: giverIrk.rawRepresentation)
    }

    /// A fresh 32-byte random nonce.
    public static func random32() -> Data {
        var b = Data(count: 32)
        b.withUnsafeMutableBytes { _ = SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
        return b
    }

    // MARK: - mailbox auth (IRK-signed DeviceEndpointClaim)

    public static func buildMailboxAuth(
        username: String,
        irk: Curve25519.Signing.PrivateKey,
        issuedAt: Int64,
        nonce: Data,
        endpointLabel: String = "device"
    ) throws -> MailboxAuthEnvelope {
        let claim = DeviceEndpointClaim(
            username: username,
            endpointLabel: endpointLabel,
            phoneIrkPub: irk.publicKey.rawRepresentation,
            issuedAt: issuedAt,
            expiresAt: issuedAt + 120_000,
            nonce: nonce
        )
        let sig = try claim.sign(with: irk)
        return MailboxAuthEnvelope(
            auth: .init(
                username: claim.username,
                endpointLabel: claim.endpointLabel,
                phoneIrkPub: HexUtil.encode(claim.phoneIrkPub),
                issuedAt: claim.issuedAt,
                expiresAt: claim.expiresAt,
                nonce: HexUtil.encode(claim.nonce)
            ),
            authSignature: HexUtil.encode(sig)
        )
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/ServerTransferFlow.kt
// Pure (JVM-testable) builders for the transfer-a-box flow
// (docs/account-deletion-and-name-reclaim.md §4). The Compose VMs derive the IRK
// behind the biometric, then call these to produce the exact wire bodies the
// broker accepts — byte-identical to the webapp lib/serverTransfer.js + iOS
// FlagshipCore/ServerTransferFlow.swift.

package com.flagshipserver.app.core

import com.flagshipserver.app.api.AdminRootTransferWire
import com.flagshipserver.app.api.MailboxAuthEnvelope
import com.flagshipserver.app.api.TransferAdminHandoffBody
import com.flagshipserver.app.api.TransferClaimBody
import com.flagshipserver.app.api.TransferClaimWire
import com.flagshipserver.app.api.TransferDiskKeyBody
import com.flagshipserver.app.api.TransferOfferBody
import com.flagshipserver.app.api.TransferOfferWire
import com.flagshipserver.app.api.TransferRehomeAuthBody
import com.google.crypto.tink.subtle.Ed25519Sign
import com.google.crypto.tink.subtle.Ed25519Verify
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.security.SecureRandom

object ServerTransferFlow {
    private val rng = SecureRandom()
    private val json = Json { encodeDefaults = true; ignoreUnknownKeys = true }

    /** The QR payload the giver shows + the acquirer scans. Field set matches the
     *  webapp `createTransferOffer().qr`. */
    @Serializable
    data class OfferQR(
        val v: Int = 1,
        val kind: String = "flagship-transfer-offer",
        val serverDomain: String,
        val transferNonce: String,
        val giverIrkPub: String,
        val issuedAt: Long,
        val expiresAt: Long,
        val offerSignature: String,
    )

    class TransferException(message: String) : RuntimeException(message)

    fun random32(): ByteArray = ByteArray(32).also { rng.nextBytes(it) }

    // ── GIVER: build the offer body + the QR ──────────────────────────────────

    data class BuiltOffer(val body: TransferOfferBody, val qr: OfferQR)

    fun buildOffer(
        serverDomain: String,
        username: String,
        irk: Ed25519Sign,
        irkPubHex: String,
        issuedAt: Long,
        // Slice D — the transfer OFFER is a SENSITIVE order: sign with the giver's
        // admin master root (`orderKey` + its `orderKeyPubHex`) when supplied, else
        // the IRK. The QR's `giverIrkPub` MUST be the key the offer was signed with
        // (the acquirer's local `verifyOfferSignature` checks `offerSignature`
        // under it), so it TRACKS the signing key. `.com` gates the offer against
        // the giver's admin root and records the giver identity from the registered
        // account, independent of this QR field. The mailbox AUTH stays IRK-signed.
        orderKey: Ed25519Sign? = null,
        orderKeyPubHex: String? = null,
        ttlMs: Long = 15 * 60_000,
        nonce: ByteArray = random32(),
        authNonce: ByteArray = random32(),
    ): BuiltOffer {
        val nonceHex = HexUtil.encode(nonce)
        val expiresAt = issuedAt + ttlMs
        val signKey = orderKey ?: irk
        val signPubHex = (if (orderKey != null) orderKeyPubHex else null) ?: irkPubHex
        val offerSig = signKey.sign(
            ServerTransferOfferOrder.canonicalBytes(serverDomain, nonceHex, issuedAt, expiresAt)
        )
        val offerSigHex = HexUtil.encode(offerSig)
        val auth = buildMailboxAuth(username, irk, irkPubHex, issuedAt, authNonce)
        val body = TransferOfferBody(
            auth = auth.auth,
            authSignature = auth.authSignature,
            offer = TransferOfferWire(serverDomain, nonceHex, issuedAt, expiresAt),
            offerSignature = offerSigHex,
        )
        val qr = OfferQR(
            serverDomain = serverDomain,
            transferNonce = nonceHex,
            giverIrkPub = signPubHex.lowercase(),
            issuedAt = issuedAt,
            expiresAt = expiresAt,
            offerSignature = offerSigHex,
        )
        return BuiltOffer(body, qr)
    }

    fun encodeQR(qr: OfferQR): String = json.encodeToString(OfferQR.serializer(), qr)

    // ── GIVER: the universal-link QR the acquirer scans ───────────────────────
    //
    // The offer JSON is a SIGNED non-secret, so it rides a base64url `o=` QUERY
    // param on the control-plane `/transfer` universal link (NOT a `#fragment` —
    // Android + the webapp strip fragments). Byte-identical across iOS/webapp so
    // any surface can scan any other's QR. The custom-scheme twin is the same
    // param on `flagship://transfer`.

    /** base64url(UTF8(offerJSON)), no padding — the `o=` param value. */
    fun encodeOfferParam(qr: OfferQR): String =
        Base64URL.encode(encodeQR(qr).toByteArray(Charsets.UTF_8))

    /** The https universal link the giver renders as a QR:
     *  `https://flagshipserver.com/transfer?o=<b64url>`. Uses the configured
     *  control apex (prod = flagshipserver.com; a gym build overrides it). */
    fun offerUrl(qr: OfferQR): String =
        "${Endpoints.controlBaseUrl}/transfer?o=${encodeOfferParam(qr)}"

    /** The custom-scheme twin: `flagship://transfer?o=<b64url>`. */
    fun offerCustomSchemeUrl(qr: OfferQR): String =
        "flagship://transfer?o=${encodeOfferParam(qr)}"

    /** Decode an `o=` param back to the offer JSON string, or null. */
    fun decodeOfferParam(param: String): String? =
        Base64URL.decode(param)?.let { runCatching { String(it, Charsets.UTF_8) }.getOrNull() }

    /** Pull the offer JSON out of a scanned/pasted string, accepting BOTH the
     *  URL forms (`…/transfer?o=<b64url>`, `flagship://transfer?o=…`) and a bare
     *  offer JSON. Pure string parse (no android.net.Uri) so it stays
     *  JVM-testable. Returns null when neither form is present. */
    fun offerJsonFrom(text: String): String? {
        val t = text.trim()
        Regex("[?&]o=([^&#]+)").find(t)?.let { return decodeOfferParam(it.groupValues[1]) }
        if (t.startsWith("{")) return t
        return null
    }

    /** Ed25519-verify the offer signature over the canonical bytes against
     *  `giverIrkPub`. A deep-linked / scanned offer is attacker-supplied, so the
     *  acquirer MUST verify this (+ expiry) BEFORE the claim biometric. Returns
     *  false on any malformed hex / bad signature (never throws). */
    fun verifyOfferSignature(qr: OfferQR): Boolean {
        val pub = HexUtil.decode(qr.giverIrkPub) ?: return false
        if (pub.size != 32) return false
        val sig = HexUtil.decode(qr.offerSignature) ?: return false
        return runCatching {
            Ed25519Verify(pub).verify(
                sig,
                ServerTransferOfferOrder.canonicalBytes(
                    qr.serverDomain, qr.transferNonce, qr.issuedAt, qr.expiresAt,
                ),
            )
            true
        }.getOrDefault(false)
    }

    // ── ACQUIRER: parse + build the claim body ────────────────────────────────

    fun parseQR(text: String): OfferQR {
        val qr = try {
            json.decodeFromString(OfferQR.serializer(), text)
        } catch (_: Throwable) {
            throw TransferException("not a transfer QR")
        }
        if (qr.kind != "flagship-transfer-offer") throw TransferException("not a transfer QR")
        if (qr.serverDomain.isEmpty() || qr.transferNonce.isEmpty() || qr.giverIrkPub.isEmpty()) {
            throw TransferException("malformed transfer QR")
        }
        return qr
    }

    fun looksLikeTransferQR(text: String): Boolean = text.contains("flagship-transfer-offer")

    fun buildClaim(
        offer: OfferQR,
        acquirerUsername: String,
        acquirerIrk: Ed25519Sign,
        acquirerIrkPubHex: String,
        issuedAt: Long,
        // Slice D — the claim's `acquirerIrkPub` field STAYS the acquirer's
        // registered IRK (`.com` requires `claim.acquirerIrkPub == acquirer
        // .irkPubHex` — identity, independent of the signature). The SENSITIVE
        // claim ORDER signs with the acquirer's admin master root (`orderKey`)
        // when supplied (else the IRK); `.com` gates the signature against the
        // acquirer's admin root.
        orderKey: Ed25519Sign? = null,
        // §9.8 — the acquirer's admin root pub rides INSIDE the v2 signed
        // canonical ("" when the account has none) so the box re-pins the
        // acquirer's anchor at re-home off a value the acquirer signed.
        acquirerAdminRootPubHex: String = "",
    ): TransferClaimBody {
        if (offer.expiresAt <= issuedAt) throw TransferException("expired")
        val lowered = acquirerUsername.lowercase()
        val sig = (orderKey ?: acquirerIrk).sign(
            ServerTransferClaimOrder.canonicalBytes(
                offer.serverDomain, offer.transferNonce, lowered, acquirerIrkPubHex,
                acquirerAdminRootPubHex, issuedAt,
            )
        )
        return TransferClaimBody(
            claim = TransferClaimWire(
                serverDomain = offer.serverDomain,
                transferNonce = offer.transferNonce,
                acquirerUsername = lowered,
                acquirerIrkPub = acquirerIrkPubHex.lowercase(),
                acquirerAdminRootPub = acquirerAdminRootPubHex.lowercase(),
                issuedAt = issuedAt,
            ),
            claimSignature = HexUtil.encode(sig),
        )
    }

    // ── GIVER: admin-root hand-off (§9.8) ─────────────────────────────────────

    /** Build + sign the admin-root hand-off the box verifies against its PINNED
     *  giver anchor before re-pinning the acquirer's root at re-home.
     *  `serverDomain` is the box's OLD canonical; `newAdminRootPubHex` "" ⇒
     *  unpin (the acquirer account has no admin root). Signed by the GIVER's
     *  admin master root. */
    fun buildAdminHandoff(
        serverDomain: String,
        giverUsername: String,
        acquirerUsername: String,
        oldAdminRootPubHex: String,
        newAdminRootPubHex: String,
        transferNonce: String,
        issuedAt: Long,
        giverAdminRoot: Ed25519Sign,
    ): TransferAdminHandoffBody {
        val handoff = AdminRootTransfer(
            serverDomain = serverDomain.lowercase(),
            giverUsername = giverUsername.lowercase(),
            acquirerUsername = acquirerUsername.lowercase(),
            oldAdminRootPub = oldAdminRootPubHex.lowercase(),
            newAdminRootPub = newAdminRootPubHex.lowercase(),
            transferNonce = transferNonce.lowercase(),
            issuedAt = issuedAt,
        )
        val sig = AdminRootTransferClaim.sign(handoff, giverAdminRoot)
        return TransferAdminHandoffBody(
            handoff = AdminRootTransferWire(
                serverDomain = handoff.serverDomain,
                giverUsername = handoff.giverUsername,
                acquirerUsername = handoff.acquirerUsername,
                oldAdminRootPub = handoff.oldAdminRootPub,
                newAdminRootPub = handoff.newAdminRootPub,
                transferNonce = handoff.transferNonce,
                issuedAt = handoff.issuedAt,
            ),
            signatureHex = HexUtil.encode(sig),
        )
    }

    // ── GIVER: legacy re-home authorization (v1-sec GAP 3) ────────────────────

    /** Build + sign the `flagship/server-rehome-auth/v1` re-home authorization
     *  with the GIVER's owner IRK (the box's pinned owner IRK until it re-homes).
     *  A box with NO pinned admin master root verifies this against its pin
     *  before writing the re-home marker — never on `.com`'s unsigned word. The
     *  deposit body carries only `issuedAt` + the signature; `.com` reconstructs
     *  the signed (old/new domain, acquirer IRK) fields from the claimed row.
     *  `oldServerDomain` is the box's OLD canonical; `newServerDomain` +
     *  `acquirerIrkPubHex` come from the giver's claim poll. */
    fun buildRehomeAuth(
        oldServerDomain: String,
        newServerDomain: String,
        acquirerIrkPubHex: String,
        giverIrk: Ed25519Sign,
        issuedAt: Long,
    ): TransferRehomeAuthBody {
        val sig = giverIrk.sign(
            RehomeAuthorizationOrder.canonicalBytes(
                oldServerDomain, newServerDomain, acquirerIrkPubHex, issuedAt,
            )
        )
        return TransferRehomeAuthBody(issuedAt = issuedAt, signatureHex = HexUtil.encode(sig))
    }

    // ── GIVER: disk-key re-seal ───────────────────────────────────────────────

    /** Re-seal the box's LUKS disk key (already unsealed by the giver IRK) to the
     *  ACQUIRER IRK pub. `.com` stays content-blind. */
    fun buildDiskKeyDeposit(
        serverDomain: String,
        username: String,
        irk: Ed25519Sign,
        irkPubHex: String,
        diskKey: ByteArray,
        acquirerIrkPubHex: String,
        issuedAt: Long,
        authNonce: ByteArray = random32(),
    ): TransferDiskKeyBody {
        val acquirerPub = HexUtil.decode(acquirerIrkPubHex)
        if (acquirerPub == null || acquirerPub.size != 32) throw TransferException("bad acquirer IRK")
        val sealed = SecretSeal.sealForEd25519Recipient(diskKey, acquirerPub)
        val auth = buildMailboxAuth(username, irk, irkPubHex, issuedAt, authNonce)
        return TransferDiskKeyBody(
            auth = auth.auth,
            authSignature = auth.authSignature,
            sealedDiskKey = HexUtil.encode(sealed),
        )
    }

    /** GIVER: open the box's install-time disk key (sealed FOR the giver IRK). */
    fun openGiverDiskKey(sealedHex: String, giverIrkSeed: ByteArray): ByteArray =
        SecretSeal.openWithEd25519Seed(
            HexUtil.decode(sealedHex) ?: throw TransferException("bad sealed key"), giverIrkSeed
        )

    /** ACQUIRER: open the giver's re-sealed disk key with the acquirer IRK seed. */
    fun openDiskKey(sealedHex: String, acquirerIrkSeed: ByteArray): ByteArray =
        SecretSeal.openWithEd25519Seed(
            HexUtil.decode(sealedHex) ?: throw TransferException("bad sealed key"), acquirerIrkSeed
        )

    // ── mailbox auth (IRK-signed DeviceEndpointClaim) ─────────────────────────

    fun buildMailboxAuth(
        username: String,
        irk: Ed25519Sign,
        irkPubHex: String,
        issuedAt: Long,
        nonce: ByteArray = random32(),
        endpointLabel: String = "device",
    ): MailboxAuthEnvelope {
        val nonceHex = HexUtil.encode(nonce)
        val expiresAt = issuedAt + 120_000
        val sig = DeviceEndpointClaim.sign(
            irk = irk,
            username = username,
            endpointLabel = endpointLabel,
            phoneIrkPubHex = irkPubHex,
            issuedAt = issuedAt,
            expiresAt = expiresAt,
            nonceHex = nonceHex,
        )
        return MailboxAuthEnvelope(
            auth = MailboxAuthEnvelope.Auth(
                username = username,
                endpointLabel = endpointLabel,
                phoneIrkPub = irkPubHex,
                issuedAt = issuedAt,
                expiresAt = expiresAt,
                nonce = nonceHex,
            ),
            authSignature = HexUtil.encode(sig),
        )
    }
}
